sdk/storage/Azure.Storage.Blobs/tests/ClientSideEncryptionTests.cs

```diff
@@ -0,0 +1,673 @@
+        }
+
+        private async Task<DisposingContainer> GetTestContainerEncryptionAsync(
+            ClientSideEncryptionOptions encryptionOptions,
+            string containerName = default,
+            IDictionary<string, string> metadata = default)
+        {
+            // normally set through property on subclass; this is easier to hook up in current test infra with internals access
+            var options = GetOptions();
+            options._clientSideEncryptionOptions = encryptionOptions;
+
+            containerName ??= GetNewContainerName();
+            var service = GetServiceClient_SharedKey(options);
+
+            BlobContainerClient container = InstrumentClient(service.GetBlobContainerClient(containerName));
+            await container.CreateAsync(metadata: metadata);
+            return new DisposingContainer(container);
+        }
+
+        private Mock<IKeyEncryptionKey> GetIKeyEncryptionKey(byte[] userKeyBytes = default, string keyId = default)
+        {
+            if (userKeyBytes == default)
+            {
+                const int keySizeBits = 256;
+                var bytes = new byte[keySizeBits >> 3];
+                new RNGCryptoServiceProvider().GetBytes(bytes);
+                userKeyBytes = bytes;
+            }
+            keyId ??= Guid.NewGuid().ToString();
+
+            var keyMock = new Mock<IKeyEncryptionKey>(MockBehavior.Strict);
+            keyMock.SetupGet(k => k.KeyId).Returns(keyId);
+            if (IsAsync)
+            {
+                keyMock.Setup(k => k.WrapKeyAsync(s_algorithmName, IsNotNull<ReadOnlyMemory<byte>>(), s_cancellationToken))
+                    .Returns<string, ReadOnlyMemory<byte>, CancellationToken>((algorithm, key, cancellationToken) => Task.FromResult(Xor(userKeyBytes, key.ToArray())));
+                keyMock.Setup(k => k.UnwrapKeyAsync(s_algorithmName, IsNotNull<ReadOnlyMemory<byte>>(), s_cancellationToken))
+                    .Returns<string, ReadOnlyMemory<byte>, CancellationToken>((algorithm, wrappedKey, cancellationToken) => Task.FromResult(Xor(userKeyBytes, wrappedKey.ToArray())));
+            }
+            else
+            {
+                keyMock.Setup(k => k.WrapKey(s_algorithmName, IsNotNull<ReadOnlyMemory<byte>>(), s_cancellationToken))
+                    .Returns<string, ReadOnlyMemory<byte>, CancellationToken>((algorithm, key, cancellationToken) => Xor(userKeyBytes, key.ToArray()));
+                keyMock.Setup(k => k.UnwrapKey(s_algorithmName, IsNotNull<ReadOnlyMemory<byte>>(), s_cancellationToken))
+                    .Returns<string, ReadOnlyMemory<byte>, CancellationToken>((algorithm, wrappedKey, cancellationToken) => Xor(userKeyBytes, wrappedKey.ToArray()));
+            }
+
+            return keyMock;
+        }
+
+        private Mock<IKeyEncryptionKeyResolver> GetAlwaysFailsKeyResolver(bool throws)
+        {
+            var mock = new Mock<IKeyEncryptionKeyResolver>(MockBehavior.Strict);
+            if (IsAsync)
```

Comment:
I'd consider removing these kind of branches and just mock both sync and async in one shot. that'd be less screens of code.
Unless we want to verify that right method was called.

Reply:
We do want to verify the right method was called. All our tests only mock the correct method and set strict mock behavior. This way, we can verify a change didn't mess up our sync/async story, which can be easy to miss.